lib/services/expiry/expiry.go

```diff
@@ -0,0 +1,250 @@
+	// Emitter is an events emitter, used to submit discrete events.
+	Emitter apievents.Emitter
+	// AccessPoint is a expiry access point.
+	AccessPoint authclient.ExpiryAccessPoint
+	// Clock is the clock used to calculate expiry.
+	Clock clockwork.Clock
+	// HostID is a unique ID of this host.
+	HostID string
+}
+
+// CheckAndSetDefaults checks required fields and sets default values.
+func (c *Config) CheckAndSetDefaults() error {
+	if c.Emitter == nil {
+		return trace.BadParameter("no Emitter configured for expiry")
+	}
+	if c.AccessPoint == nil {
+		return trace.BadParameter("no AccessPoint configured for expiry")
+	}
+	if c.Clock == nil {
+		return trace.BadParameter("no Clock configured for expiry")
+	}
+	return nil
+}
+
+// Service is a expiry service.
+type Service struct {
+	*Config
+
+	ctx      context.Context
+	cancelfn context.CancelFunc
+}
+
+// New initializes a expiry service
+func New(ctx context.Context, cfg *Config) (*Service, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	s := &Service{
+		Config: cfg,
+		ctx:    ctx,
+	}
+	return s, nil
+}
+
+// Run starts the expiry service.
+func (s *Service) Run() error {
+	semCfg := services.SemaphoreLockConfigWithRetry{
+		SemaphoreLockConfig: services.SemaphoreLockConfig{
+			Service: s.AccessPoint,
+			Params: types.AcquireSemaphoreRequest{
+				SemaphoreKind: types.KindAccessRequest,
+				SemaphoreName: semaphoreName,
+				MaxLeases:     1,
+				Holder:        s.HostID,
+			},
+			Expiry: semaphoreExpiration,
+			Clock:  s.Clock,
+		},
+		Retry: retryutils.LinearConfig{
+			Clock:  s.Clock,
+			First:  time.Second,
+			Step:   semaphoreExpiration / 2,
+			Max:    semaphoreExpiration,
+			Jitter: retryutils.DefaultJitter,
+		},
+	}
+
+	poll := interval.New(interval.Config{
+		Duration:      scanInterval,
+		FirstDuration: retryutils.FullJitter(scanInterval),
+		Jitter:        retryutils.SeventhJitter,
+		Clock:         s.Clock,
+	})
+	defer poll.Stop()
+
+	for {
+		lease, err := services.AcquireSemaphoreLockWithRetry(
+			s.ctx,
+			semCfg,
+		)
+		if err != nil {
+			s.Log.WarnContext(s.ctx, "error aquiring semaphore", "error", err)
+			continue
+		}
+
+		if err := s.processRequests(); err != nil {
+			s.Log.WarnContext(s.ctx, "error processing access requests", "error", err)
+		}
+		ctx, cancel := context.WithCancel(lease)
+		defer cancel()
+		lease.Stop()
+		if err := lease.Wait(); err != nil {
+			s.Log.WarnContext(ctx, "error cleaning up semaphore", "error", err)
+		}
+		select {
+		case <-s.ctx.Done():
+			return nil
+		case <-poll.Next():
+		}
+	}
+}
+func (s *Service) processRequests() error {
+	requestsExpired := 0
+	nextPageStart := ""
+	for {
+		var page []*types.AccessRequestV3
+		var err error
+		readTime := s.Clock.Now()
+		page, nextPageStart, err = s.getNextPageOfAccessRequests(nextPageStart)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if len(page) == 0 || requestsExpired >= maxExpiresPerCycle {
```

Comment:
Isn't `requestsExpired >= maxExpiresPerCycle` condition here redundant? Wouldn't we exit below on line 184 if this is true?

I think we should remove it either below or from here.

Reply:
Removed the first one.